docs: fix description of treemanager api

Theeraphat-Sorasetsakul opened the pull request.

Commit:
docs: fix description of treemanager api

Theeraphat-Sorasetsakul committed:
Merge branch 'feat/tree-node/main' into feat/tree-node/doc

Commit:
docs: add tree-manager page

Commit:
Merge branch 'feat/tree-node/doc' of https://github.com/Refinitiv/refinitiv-ui into feat/tree-node/doc

Commit:
docs: add isReadonly stamp on doc

Commit:
docs: improve generate doc to support readonly for accessor

Commit:
Merge remote-tracking branch 'origin/feat/tree-node/main' into feat/tree-node/doc

Commit:
docs: add manipulate section to tree and tree-select

Theeraphat-Sorasetsakul committed:
Merge branch 'feat/tree-node/main' into feat/tree-node/doc

wattachai-lseg commented on packages/elements/src/tree/managers/tree-manager.ts:
```diff
@@ -107,7 +107,7 @@ export class TreeManager<T extends TreeDataItem> {
 
   /**
-   * Returns all checked items.
+   * Returns selected items which selected property is true.
```
```suggestion
   * Returns all selected items.
```

wattachai-lseg commented on packages/elements/src/tree/managers/tree-manager.ts:
```diff
@@ -134,8 +134,8 @@ export class TreeManager<T extends TreeDataItem> {
-   * Items which should be visibly displayed.
-   * This can be used to render items.
+   * Returns items which selected prop can be changed.
+   * The items don't include hided, disabled, readonly.
```
```suggestion
   * Returns items which their selected state can be changed.
   * hidden, disabled or readonly items are not included.
```

wattachai-lseg commented on packages/elements/src/tree/managers/tree-manager.ts:
```diff
@@ -160,8 +160,8 @@ export class TreeManager<T extends TreeDataItem> {
-   * Items which should be visibly displayed.
-   * This can be used to render items.
+   * Returns current displayed items.
+   * The items don't include hided, and children of unexpanded parent.
```
```suggestion
   * Returns currently displayed items.
   * hidden and children of unexpanded items are not included.
```

wattachai-lseg commented on packages/elements/src/tree/managers/tree-manager.ts:
```diff
@@ -284,7 +284,7 @@ export class TreeManager<T extends TreeDataItem> {
 
   /**
-   * Updates the data item, forcing a render
+   * Call to update the data item and emit event to render
```
> Request the item to be re-rendered manually. This should be unnecessary most of the time as re-render is internally triggered by property update already

taken from `feat/tree-node/docs-tree-node`

Theeraphat-Sorasetsakul replied:
Updated.

wattachai-lseg commented on packages/elements/src/tree/managers/tree-manager.ts:
```diff
@@ -305,6 +305,7 @@ export class TreeManager<T extends TreeDataItem> {
 
   /**
+   * Set hidden property of the item to true.
    * Excludes an item as part of the tree.
```
remove this line?

Theeraphat-Sorasetsakul replied:
Removed.

wattachai-lseg commented on packages/elements/src/tree/managers/tree-manager.ts:
```diff
@@ -316,7 +317,7 @@ export class TreeManager<T extends TreeDataItem> {
 
   /**
-   * Is the item checkable?
+   * Return if the selected property of item can be changed.
```
```suggestion
   * Return if the selected state of item can be changed.
```

wattachai-lseg commented on packages/elements/src/tree/managers/tree-manager.ts:
```diff
@@ -329,7 +330,7 @@ export class TreeManager<T extends TreeDataItem> {
 
   /**
-   * Is the item expanded?
+   * Return if the expanded property of item is true.
```
```suggestion
   * Return the current expanded state of the item.
```

wattachai-lseg commented on scripts/release/class-api-analyzer.js:
```diff
@@ -6,6 +6,27 @@ import TypeDoc from 'typedoc';
+  const classIDs = data.groups.find((item) => item.title === 'Classes')?.children;
+  classIDs.forEach((classID) => {
+    const { groups, children } = data.children.find((itemClass) => itemClass.id === classID);
+    groups
```
let's make it easier to read by creating a variable called `accessorIDs` first.

Theeraphat-Sorasetsakul replied:
Fixed.

wattachai-lseg commented on scripts/release/class-api-analyzer.js:
```diff
@@ -6,6 +6,27 @@ import TypeDoc from 'typedoc';
+        const accessorItem = children.find((accessorItem) => accessorID === accessorItem.id);
+        if (!accessorItem.setSignature) {
+          accessorItem.flags.isReadonly = true;
+          accessorItem.getSignature.flags.isReadonly = true;
```
Is there a recommended method for custom flags provided by TypeDoc? Let's do our best to be forward compatible here.

Theeraphat-Sorasetsakul replied:
They suggest checking the flag at reflection.setSignature and reflection.getSignature. I tried to get there but it doesn't be straightforward to get the reflection. So, I decide stamping it by our side. 
P.S. reflection is the type of Typedoc. reflection.setSignature is setter accessor.

wattachai-lseg commented on scripts/release/utils.injector.js:
```diff
@@ -165,7 +177,8 @@ const generateMethod = (methodIDs, dataClass, mappedSignatures) => {
     for (const signature of data.signatures) {
-      result.push({ h3: signature.name });
+      const header = signature.flags.isReadonly ? `${signature.name} <code>readonly</code>` : signature.name;
+      result.push({ h3: header });
```
refactor into a readonly header util?

Theeraphat-Sorasetsakul replied:
Rollbacked as we discussed to put readonly as next section.

wattachai-lseg commented on documents/src/pages/elements/tree-select.md:
```diff
@@ -520,6 +520,24 @@ This control will save the current selection, and result of the firing of the `v
 
 This control will abort any current selection changes and go back to the tree state when first opened.
 
+## Manipulate item property
```
```suggestion
## Manipulating item properties
```

wattachai-lseg commented on documents/src/pages/elements/tree-select.md:
````diff
@@ -520,6 +520,24 @@ This control will save the current selection, and result of the firing of the `v
+Tree Select component supports functionality for updating and reading data to manipulate items properties. These are used in conjunction with [TreeManager](./custom-components/utils/tree-manager) from `treeManager` property and [TreeNode](./custom-components/utils/tree-node) which can be retrieved from `getTreeNode` through `treeManager` property.
+
+```javascript
+const treeSelect = document.querySelector('ef-tree-select');
````
let's provide a more concrete example. Check Tree Node doc in `feat/tree-node/docs-tree-node` for a real example.

Theeraphat-Sorasetsakul replied:
Added.

wattachai-lseg commented on documents/src/pages/elements/tree-select.md:
```diff
@@ -520,6 +520,24 @@ This control will save the current selection, and result of the firing of the `v
 
+## Manipulate item property
+
+Tree Select component supports functionality for updating and reading data to manipulate items properties. These are used in conjunction with [TreeManager](./custom-components/utils/tree-manager) from `treeManager` property and [TreeNode](./custom-components/utils/tree-node) which can be retrieved from `getTreeNode` through `treeManager` property.
```
```suggestion
Item properties of Tree Select could be read and updated programmatically through its [TreeManager](./custom-components/utils/tree-manager) which is available as `treeManager` property. Retrieve [TreeNode(s)](./custom-components/utils/tree-node) representing each item by calling `getTreeNode()` or `getTreeNodes()` of `treeManager`. 
```

wattachai-lseg commented on documents/src/pages/elements/tree-select.md:
```diff
@@ -520,6 +520,24 @@ This control will save the current selection, and result of the firing of the `v
+if (treeSelect) {
+  const node0 = treeSelect.treeManager.getTreeNodes()[0]; // get the item at index number 0.
+  node0.label = 'new label'; // edit label property which this will reflect to the component.
+}
```
non-null assertion should be typescript example easier to read.

```suggestion
const treeSelect = document.querySelector('ef-tree-select')!;
const node0 = treeSelect.treeManager.getTreeNodes()[0]; // get the item at index number 0.
node0.label = 'new label'; // edit label property which this will reflect to the component.
```

wattachai-lseg commented on documents/src/pages/elements/tree.md:
````diff
@@ -455,6 +455,24 @@ tree?.addEventListener('value-changed', (event) => {
 });
 ```
 
+## Manipulate item property
````
Let's refer to Tree Select's feedback.

Theeraphat-Sorasetsakul replied:
Fixed.

wattachai-lseg commented on packages/elements/src/tree/managers/tree-manager.ts:
* constructor: Let's mention that usually there should be no need to call it directly. Use the one exposed by the component instead.
* Could you try to apply the same principle of the feedbacks to all APIs?

Theeraphat-Sorasetsakul replied:
Applied to all APIs.

wattachai-lseg commented on documents/src/pages/utils/tree-manager.md:
```diff
@@ -0,0 +1,61 @@
+
+# TreeManager
+
+Tree Manager is a simplified version of Collection Composer class that provide an ability to manage the nested items in [Tree](/elements/tree)/[Tree Select](/elements/tree-select) component. The class has an API for reading and updating data to access item properties. This works in conjunction with [Tree Node](/custom-components/utils/tree-node). 
```
Shall we also mention the following? 
> Most of the time, getTreeNodes(), getTreeNode() and TreeNode are all you need. The remaining APIs are either used internally or maintained for backward compatibility.

Theeraphat-Sorasetsakul replied:
Updated.

wattachai-lseg commented on documents/src/pages/utils/tree-manager.md:
````diff
@@ -0,0 +1,61 @@
+Tree Manager is a simplified version of Collection Composer class that provide an ability to manage the nested items in [Tree](/elements/tree)/[Tree Select](/elements/tree-select) component. The class has an API for reading and updating data to access item properties. This works in conjunction with [Tree Node](/custom-components/utils/tree-node). 
+
+::
+```javascript
````
I'm not sure that the live example is needed here...

Theeraphat-Sorasetsakul replied:
Updated.

wattachai-lseg commented on documents/src/pages/utils/tree-manager.md:
```diff
@@ -0,0 +1,61 @@
+
+# TreeManager
+
+Tree Manager is a simplified version of Collection Composer class that provide an ability to manage the nested items in [Tree](/elements/tree)/[Tree Select](/elements/tree-select) component. The class has an API for reading and updating data to access item properties. This works in conjunction with [Tree Node](/custom-components/utils/tree-node). 
```
Let's describe about type param/generic support so our readers would understand why some APIs return `T[]`.

Theeraphat-Sorasetsakul replied:
Updated.

wattachai-lseg commented on documents/src/pages/utils/tree-manager.md:
* unexpected readonly in current-page navigation menu
![image](https://github.com/Refinitiv/refinitiv-ui/assets/117723407/9d510431-f1ec-446e-96a8-3eae2cee1f81)

Theeraphat-Sorasetsakul replied:
Fixed to put readonly to next section next to heading as we discussed yesterday.

Commit:
fix: move readonly to next section next to heading

Theeraphat-Sorasetsakul committed:
Apply suggestions from code review

Co-authored-by: wattachai <117723407+wattachai-lseg@users.noreply.github.com>

wattachai-lseg commented on scripts/release/utils.injector.js:
```diff
@@ -166,6 +176,7 @@ const generateMethod = (methodIDs, dataClass, mappedSignatures) => {
 
     for (const signature of data.signatures) {
       result.push({ h3: signature.name });
+      signature.flags.isReadonly && result.push({ p: '<code>readonly</code>' });
```
methods are expected to be called not get or set. Hence, read-only doesn't apply here, right?

Theeraphat-Sorasetsakul replied:
Yes, but I put this for supporting if the JSDoc of method has @readonly.

wattachai-lseg replied:
yeah but it still doesn't make sense for methods. Shall we remove it?

Theeraphat-Sorasetsakul replied:
I think we can keep it because it's an option to document and it's feature from JSDoc.

wattachai-lseg replied:
I still insist that we should avoid adding unnecessary code.

Theeraphat-Sorasetsakul replied:
OK let remove it.

wattachai-lseg commented on documents/src/pages/utils/tree-manager.md:
```diff
@@ -0,0 +1,61 @@
+layout: default
+-->
+
+# TreeManager
```
```suggestion
# Tree Manager
```

wattachai-lseg commented on documents/src/pages/utils/tree-manager.md:
```diff
@@ -0,0 +1,61 @@
+<!-- 
+title: TreeManager
```
```suggestion
title: Tree Manager
```

Commit:
fix: do stamping isReadonly at mappedSignature

Commit:
fix: fix and add example to tree-select and tree doc

Theeraphat-Sorasetsakul commented on packages/elements/src/tree/managers/tree-manager.ts:
```diff
@@ -305,6 +305,7 @@ export class TreeManager<T extends TreeDataItem> {
 
   /**
+   * Set hidden property of the item to true.
    * Excludes an item as part of the tree.
```
```suggestion
```

Theeraphat-Sorasetsakul committed:
Apply suggestions from code review

Co-authored-by: wattachai <117723407+wattachai-lseg@users.noreply.github.com>

Commit:
fix: fix and update doc

Commit:
fix: improve example in Tree Manager doc

Commit:
fix: add support language tabs

Commit:
Merge branch 'feat/tree-node/doc' of https://github.com/Refinitiv/refinitiv-ui into feat/tree-node/doc

wattachai-lseg commented on scripts/release/class-api-analyzer.js:
```diff
@@ -6,6 +6,31 @@ import TypeDoc from 'typedoc';
+ * @param {object} data json data generated by Typedoc
+ * @returns {void}
+ */
+const stampIsReadonly = (data) => {
```
```suggestion
const flagReadonlyAccessors = (data) => {
```

wattachai-lseg commented on packages/elements/src/tree/managers/tree-manager.ts:
```diff
@@ -40,6 +40,8 @@ export class TreeManager<T extends TreeDataItem> {
   private treeNodeCache = new Map<T, TreeNode<T>>();
 
   /**
+   * A special method to be called when initializes the class.
```
it seems this jsdoc comment doesn't get picked up by TypeDoc. I don't think that the comment is important enough to justify the need of scripting enhancement. Let's remove it instead.

Theeraphat-Sorasetsakul replied:
Updated to fix to display description on doc and rewrite the hole sentences.

wattachai-lseg commented on packages/elements/src/tree/managers/tree-manager.ts:
```diff
@@ -100,14 +102,14 @@ export class TreeManager<T extends TreeDataItem> {
 
   /**
-   * Return all items which have children
+   * Returns all items which have children
```
```suggestion
   * Returns all items with children
```

wattachai-lseg commented on packages/elements/src/tree/managers/tree-manager.ts:
```diff
@@ -66,7 +68,7 @@ export class TreeManager<T extends TreeDataItem> {
+   * Returns a `TreeNode` of the specified item.
    * If the item doesn't exist, return `null`.
    * @param item specified item
    * @returns `TreeNode` or `null`
```
```suggestion
   * @returns `TreeNode` of the original data item or `null`
```

wattachai-lseg commented on packages/elements/src/tree/managers/tree-manager.ts:
```diff
@@ -49,7 +51,7 @@ export class TreeManager<T extends TreeDataItem> {
   /**
-   * Return all items as an array of `TreeNode`.
+   * Returns all items as an array of `TreeNode`.
    * @returns  array of `TreeNode`
```
```suggestion
   * @returns Array of `TreeNode` representing all items
```

wattachai-lseg commented on packages/elements/src/tree/managers/tree-manager.ts:
```diff
@@ -284,7 +286,8 @@ export class TreeManager<T extends TreeDataItem> {
 
   /**
-   * Updates the data item, forcing a render
+   * Re-render the item manually.
+   * However, most of the time as re-render is internally triggered by property update already.
```
```suggestion
   * Requests the item to be rerendered manually.
   * This should be unnecessary most of the time as rerender is internally triggered by property update already.
```

wattachai-lseg commented on packages/elements/src/tree/managers/tree-manager.ts:
```diff
@@ -66,7 +68,7 @@ export class TreeManager<T extends TreeDataItem> {
-   * Return a `TreeNode` of the specified item.
+   * Returns a `TreeNode` of the specified item.
    * If the item doesn't exist, return `null`.
    * @param item specified item
```
```suggestion
   * @param item Original data item
```

wattachai-lseg commented on packages/elements/src/tree/managers/tree-manager.ts:
```diff
@@ -66,7 +68,7 @@ export class TreeManager<T extends TreeDataItem> {
 
   /**
-   * Return a `TreeNode` of the specified item.
+   * Returns a `TreeNode` of the specified item.
```
```suggestion
   * Returns a `TreeNode` of the original data item.
```

wattachai-lseg commented on packages/elements/src/tree/managers/tree-manager.ts:
```diff
@@ -293,7 +296,7 @@ export class TreeManager<T extends TreeDataItem> {
-   * Includes an item as part of the tree.
+   * Shows the item to be displayed.
    * @hidden `hidden` usage in filterItems of Tree & Tree Select component conflicts with this API
    * @param item Item to include
```
```suggestion
   * @param item Original data item
```

wattachai-lseg commented on packages/elements/src/tree/managers/tree-manager.ts:
```diff
@@ -305,7 +308,7 @@ export class TreeManager<T extends TreeDataItem> {
-   * Excludes an item as part of the tree.
+   * Hides the item to be unrevealed.
    * @hidden `hidden` usage in filterItems of Tree & Tree Select component conflicts with this API
    * @param item Item to exclude
```
```suggestion
   * @param item Original data item
```

wattachai-lseg commented on packages/elements/src/tree/managers/tree-manager.ts:
```diff
@@ -293,7 +296,7 @@ export class TreeManager<T extends TreeDataItem> {
 
   /**
-   * Includes an item as part of the tree.
+   * Shows the item to be displayed.
```
```suggestion
   * Shows the item.
```

wattachai-lseg commented on packages/elements/src/tree/managers/tree-manager.ts:
```diff
@@ -305,7 +308,7 @@ export class TreeManager<T extends TreeDataItem> {
 
   /**
-   * Excludes an item as part of the tree.
+   * Hides the item to be unrevealed.
```
```suggestion
   * Hides the item.
```

wattachai-lseg commented on packages/elements/src/tree/managers/tree-manager.ts:
```diff
@@ -316,7 +319,7 @@ export class TreeManager<T extends TreeDataItem> {
 
   /**
-   * Is the item checkable?
+   * Returns if the selected state of item can be changed.
```
```suggestion
   * Returns whether the selected state of item can be changed or not.
```

wattachai-lseg commented on packages/elements/src/tree/managers/tree-manager.ts:
```diff
@@ -338,7 +341,7 @@ export class TreeManager<T extends TreeDataItem> {
 
   /**
-   * Is the item a parent?
+   * Returns if the item contains the child items.
```
```suggestion
   * Returns whether the item contains any children or not.
```

wattachai-lseg commented on packages/elements/src/tree/managers/tree-manager.ts:
```diff
@@ -347,7 +350,7 @@ export class TreeManager<T extends TreeDataItem> {
 
   /**
-   * Is the item a child?
+   * Returns if the item direct to the another item.
```
```suggestion
   * Returns whether the item has a parent or not.
```

wattachai-lseg commented on packages/elements/src/tree/managers/tree-manager.ts:
```diff
@@ -356,9 +359,9 @@ export class TreeManager<T extends TreeDataItem> {
+   * Returns state which is 1=selected, 0=deselected or -1=intermediate.
    * @param item Original data item
-   * @returns Checked state of the item
+   * @returns Selected state of the item
```
```suggestion
   * @returns item checked state: `1` for checked, `0` for unchecked and `-1` for indeterminate
```

wattachai-lseg commented on packages/elements/src/tree/managers/tree-manager.ts:
```diff
@@ -356,9 +359,9 @@ export class TreeManager<T extends TreeDataItem> {
 
   /**
-   * Calculates the checked stated of the item
+   * Returns state which is 1=selected, 0=deselected or -1=intermediate.
```
```suggestion
   * Return checked state of the item.
```

wattachai-lseg commented on packages/elements/src/tree/managers/tree-manager.ts:
```diff
@@ -371,7 +374,7 @@ export class TreeManager<T extends TreeDataItem> {
 
   /**
-   * Gets an item's ancestors
+   * Returns all parent items of the item.
```
```suggestion
   * Returns all ancestors of the item.
```

wattachai-lseg commented on packages/elements/src/tree/managers/tree-manager.ts:
```diff
@@ -380,7 +383,7 @@ export class TreeManager<T extends TreeDataItem> {
 
   /**
-   * Gets an item's descendants
+   * Returns all child items of the item.
```
```suggestion
   * Returns all descendants of the item.
```

wattachai-lseg commented on packages/elements/src/tree/managers/tree-manager.ts:
```diff
@@ -390,7 +393,7 @@ export class TreeManager<T extends TreeDataItem> {
 
   /**
-   * Gets an item's parent, if it has one.
+   * Returns the one upper level parents of the item.
```
```suggestion
   * Returns the parent of the item, if it has one.
```

wattachai-lseg commented on packages/elements/src/tree/managers/tree-manager.ts:
```diff
@@ -399,7 +402,7 @@ export class TreeManager<T extends TreeDataItem> {
 
   /**
-   * Gets an item's child collection
+   * Returns the first depth children of the item.
```
```suggestion
   * Returns the children of the item as an array.
```

wattachai-lseg commented on packages/elements/src/tree/managers/tree-manager.ts:
```diff
@@ -399,7 +402,7 @@ export class TreeManager<T extends TreeDataItem> {
-   * Gets an item's child collection
+   * Returns the first depth children of the item.
    * @param item Original data item
    * @returns A list of children
```
```suggestion
   * @returns An array of children
```

wattachai-lseg commented on packages/elements/src/tree/managers/tree-manager.ts:
```diff
@@ -408,7 +411,7 @@ export class TreeManager<T extends TreeDataItem> {
 
   /**
-   * Expand an item to show its children
+   * Expands the item to show its direct children.
```
```suggestion
   * Expands the item to show its children.
```

wattachai-lseg commented on packages/elements/src/tree/managers/tree-manager.ts:
```diff
@@ -430,23 +433,23 @@ export class TreeManager<T extends TreeDataItem> {
 
   /**
-   * Expands all items in the tree
+   * Expands all items to show all child items.
```
"show items" is not just abot `expand`, it's about `hidden` as well.
```suggestion
   * Expands all items managed by the Tree Manager
```

wattachai-lseg commented on packages/elements/src/tree/managers/tree-manager.ts:
```diff
@@ -430,23 +433,23 @@ export class TreeManager<T extends TreeDataItem> {
 
   /**
-   * Collapses all items in the tree
+   * Collapses all items to hide all  child items.
```
"show items" is not just abot `expand`, it's about `hidden` as well.
```suggestion
   * Collapses all items managed by the Tree Manager
```

wattachai-lseg commented on packages/elements/src/tree/managers/tree-manager.ts:
```diff
@@ -494,7 +497,7 @@ export class TreeManager<T extends TreeDataItem> {
 
   /**
-   * Toggles the item checked state
+   * Switch selected state of the item.
```
```suggestion
   * Toggle the selected state of the item.
```

wattachai-lseg commented on documents/src/pages/utils/tree-manager.md:
```diff
@@ -0,0 +1,94 @@
+
+Tree Manager is a simplified version of Collection Composer class that provide an ability to manage the nested items in [Tree](/elements/tree)/[Tree Select](/elements/tree-select) component. The class has an API for reading and updating data to access item properties. This works in conjunction with [Tree Node](/custom-components/utils/tree-node). Most of the time, getTreeNodes(), getTreeNode() an ...
+
+Tree manager supports for generic type which allow users to define the type of item. In this document, the generic type is noted as `T`.
```
```suggestion
@> Tree Manager supports type generic defining the type of its items. In this document, this generic type is noted as `T`.
```

wattachai-lseg commented on documents/src/pages/utils/tree-manager.md:
```diff
@@ -0,0 +1,94 @@
+
+# Tree Manager
+
+Tree Manager is a simplified version of Collection Composer class that provide an ability to manage the nested items in [Tree](/elements/tree)/[Tree Select](/elements/tree-select) component. The class has an API for reading and updating data to access item properties. This works in conjunction with [Tree Node](/custom-components/utils/tree-node). Most of the time, getTreeNodes(), getTreeNode() an ...
```
Let's add link to Collection Composer page.

wattachai-lseg replied:
```suggestion
Tree Manager is a simplified version of Collection Composer providing an ability to manage nested items in [Tree](/elements/tree)/[Tree Select](/elements/tree-select) component with APIs for reading and updating item properties. It works in conjunction with [Tree Node](/custom-components/utils/tree-node). Most of the time, getTreeNodes(), getTreeNode() and TreeNode are all you need.
```

wattachai-lseg commented on documents/src/pages/utils/tree-manager.md:
````diff
@@ -0,0 +1,94 @@
+
+Tree manager supports for generic type which allow users to define the type of item. In this document, the generic type is noted as `T`.
+
+```javascript
````
With these [Tree Node code examples](https://github.com/Refinitiv/refinitiv-ui/pull/1145/files), I don't think we need another one here. Besides, this example uses `checkedItems()` while we just mentioned earlier that treeNode related APIs are all you need...

Theeraphat-Sorasetsakul replied:
Removed.

wattachai-lseg commented on documents/src/pages/elements/tree.md:
```diff
@@ -455,6 +455,50 @@ tree?.addEventListener('value-changed', (event) => {
 
+## Manipulating item properties
+
+Item properties of Tree could be read and updated programmatically through its [TreeManager](./custom-components/utils/tree-manager) which is available as `manager` property. Retrieve [TreeNode(s)](./custom-components/utils/tree-node) representing each item by calling `getTreeNode()` or `getTreeNodes()` of `manager`. 
```
Let's cleanup trailing whitespaces.

wattachai-lseg replied:
```suggestion
Item properties of Tree could be read and updated programmatically through its [TreeManager](./custom-components/utils/tree-manager) which is available as `manager` property. Retrieve [TreeNode(s)](./custom-components/utils/tree-node) representing each item by calling `getTreeNode()` or `getTreeNodes()` of `manager`. In adiition, they could be used for [custom renderer](./custom-components/utils/tree-node#custom-renderer) too. 
```
```

wattachai-lseg commented on documents/src/pages/elements/tree.md:
```diff
@@ -455,6 +455,50 @@ tree?.addEventListener('value-changed', (event) => {
+    }]
+  }
+];
+const node0 = tree.manager.getTreeNodes()[0]; // get the item at index number 0.
```
naming variable with numbers looks sloppy.
```suggestion
const firstNode = tree.manager.getTreeNodes()[0]; // get the item at index number 0.
```

wattachai-lseg commented on documents/src/pages/elements/tree.md:
```diff
@@ -455,6 +455,50 @@ tree?.addEventListener('value-changed', (event) => {
+    }]
+  }
+];
+const node0 = tree.manager.getTreeNodes()[0]; // get the item at index number 0.
```
naming variable with numbers looks sloppy.
```suggestion
const firstNode = tree.manager.getTreeNodes()[0]; // get the item at index number 0.
```

wattachai-lseg commented on documents/src/pages/elements/tree-select.md:
```diff
@@ -520,6 +520,48 @@ This control will save the current selection, and result of the firing of the `v
 
+## Manipulating item properties
+
+Item properties of Tree Select could be read and updated programmatically through its [TreeManager](./custom-components/utils/tree-manager) which is available as `treeManager` property. Retrieve [TreeNode(s)](./custom-components/utils/tree-node) representing each item by calling `getTreeNode()` or `getTreeNodes()` of `treeManager`. 
```
let's cleanup trailing whitespaces.

wattachai-lseg replied:
```suggestion
Item properties of Tree Select could be read and updated programmatically through its [TreeManager](./custom-components/utils/tree-manager) which is available as `treeManager` property. Retrieve [TreeNode(s)](./custom-components/utils/tree-node) representing each item by calling `getTreeNode()` or `getTreeNodes()` of `treeManager`. In adiition, they could be used for [custom renderer](./custom-components/utils/tree-node#custom-renderer) too. 
```

wattachai-lseg commented on documents/src/pages/elements/tree-select.md:
```diff
@@ -520,6 +520,48 @@ This control will save the current selection, and result of the firing of the `v
+    }]
+  }
+];
+const node0 = treeSelect.treeManager.getTreeNodes()[0]; // get the item at index number 0.
```
naming variable with numbers looks sloppy.

Theeraphat-Sorasetsakul replied:
Fixed.

wattachai-lseg commented on documents/src/pages/elements/tree-select.md:
```diff
@@ -520,6 +520,48 @@ This control will save the current selection, and result of the firing of the `v
+    }]
+  }
+];
+const node0 = treeSelect.treeManager.getTreeNodes()[0]; // get the item at index number 0.
```
naming variable with numbers looks sloppy.

Theeraphat-Sorasetsakul replied:
Fixed.

wattachai-lseg commented on documents/src/pages/elements/tree.md:
````diff
@@ -455,6 +455,50 @@ tree?.addEventListener('value-changed', (event) => {
+```
+
+```typescript
+const tree = document.querySelector('ef-tree')!;    
````
let's cleanup trailing whitespaces.

Theeraphat-Sorasetsakul replied:
Updated.

wattachai-lseg committed:
Merge branch 'feat/tree-node/main' into feat/tree-node/doc

Theeraphat-Sorasetsakul commented on documents/src/pages/elements/tree-select.md:
```diff
@@ -520,6 +520,48 @@ This control will save the current selection, and result of the firing of the `v
+    }]
+  }
+];
+const node0 = treeSelect.treeManager.getTreeNodes()[0]; // get the item at index number 0.
```
```suggestion
const firstNode = treeSelect.treeManager.getTreeNodes()[0]; // get the item at index number 0.
```

Theeraphat-Sorasetsakul commented on documents/src/pages/elements/tree-select.md:
```diff
@@ -520,6 +520,48 @@ This control will save the current selection, and result of the firing of the `v
+    }]
+  }
+];
+const node0 = treeSelect.treeManager.getTreeNodes()[0]; // get the item at index number 0.
```
```suggestion
const firstNode = treeSelect.treeManager.getTreeNodes()[0]; // get the item at index number 0.
```

Theeraphat-Sorasetsakul committed:
Apply suggestions from code review

Co-authored-by: wattachai <117723407+wattachai-lseg@users.noreply.github.com>

Theeraphat-Sorasetsakul committed:
Apply suggestions from code review

Co-authored-by: wattachai <117723407+wattachai-lseg@users.noreply.github.com>

Commit:
fix: remove unused space and clean doc

Theeraphat-Sorasetsakul committed:
Apply suggestions from code review

Co-authored-by: wattachai <117723407+wattachai-lseg@users.noreply.github.com>

Commit:
fix: improve doc

wattachai-lseg commented on scripts/release/utils.injector.js:
```diff
@@ -259,12 +273,14 @@ const generateMD = () => {
 
       const outputFile = path.resolve(Build.PAGES_FOLDER, output);
       if (fs.existsSync(outputFile)) {
+        // add new line to syntax error with exist content.
+        markdown = '\n' + markdown;
```
the comment doesn't look right.

Theeraphat-Sorasetsakul replied:
Fixed to be "add new line to split with exist content."

wattachai-lseg commented on scripts/release/utils.injector.js:
```diff
@@ -165,6 +178,7 @@ const generateMethod = (methodIDs, dataClass, mappedSignatures) => {
     }
 
     for (const signature of data.signatures) {
+      const mappedSignature = mappedSignatures.find((item) => item.id === signature.id);
```
move the expression to line 196 instead without assigning it to a variable?

Theeraphat-Sorasetsakul replied:
Fixed.

Theeraphat-Sorasetsakul commented on scripts/release/utils.injector.js:
```diff
@@ -259,12 +273,14 @@ const generateMD = () => {
 
       const outputFile = path.resolve(Build.PAGES_FOLDER, output);
       if (fs.existsSync(outputFile)) {
+        // add new line to syntax error with exist content.
```
```suggestion
        // add new line to split with exist content.
```

Theeraphat-Sorasetsakul replied:
```suggestion
        // add new line to split with existรืเ content.
```

Theeraphat-Sorasetsakul committed:
Update scripts/release/utils.injector.js

wattachai-lseg commented on packages/elements/src/tree/managers/tree-manager.ts:
```diff
@@ -40,6 +40,8 @@ export class TreeManager<T extends TreeDataItem> {
   private treeNodeCache = new Map<T, TreeNode<T>>();
 
   /**
+   * It is not recommended to create this instance manually.
+   * Use the existing instance in component instead.
```
```suggestion
   * Most of the time, there is no need to create a new instance of Tree Manager manually.
   * Use the existing instance in components instead.
```

Theeraphat-Sorasetsakul committed:
Update packages/elements/src/tree/managers/tree-manager.ts

Co-authored-by: wattachai <117723407+wattachai-lseg@users.noreply.github.com>

wattachai-lseg commented on packages/elements/src/tree/managers/tree-manager.ts:
```diff
@@ -430,23 +433,23 @@ export class TreeManager<T extends TreeDataItem> {
 
   /**
-   * Expands all items in the tree
+   * Expands all items managed by the Tree Manager
```
```suggestion
   * Expands all items managed by the Tree Manager.
```

wattachai-lseg commented on packages/elements/src/tree/managers/tree-manager.ts:
```diff
@@ -430,23 +433,23 @@ export class TreeManager<T extends TreeDataItem> {
 
   /**
-   * Collapses all items in the tree
+   * Collapses all items managed by the Tree Manager
```
```suggestion
   * Collapses all items managed by the Tree Manager.
```

wattachai-lseg commented on documents/src/pages/elements/tree.md:
```diff
@@ -455,6 +455,50 @@ tree?.addEventListener('value-changed', (event) => {
 
+## Manipulating item properties
+
+Item properties of Tree could be read and updated programmatically through its [TreeManager](./custom-components/utils/tree-manager) which is available as `manager` property. Retrieve [TreeNode(s)](./custom-components/utils/tree-node) representing each item by calling `getTreeNode()` or `getTreeNodes()` of `manager`. In adiition, they could be used for [custom renderer](./custom-components/utils/ ...
```
```suggestion
Item properties of Tree could be read and updated programmatically through its [TreeManager](./custom-components/utils/tree-manager) which is available as `manager` property. Retrieve [TreeNode(s)](./custom-components/utils/tree-node) representing each item by calling `getTreeNode()` or `getTreeNodes()` of `manager`. In addition, they could be used for [custom renderer](./custom-components/utils/tree-node#custom-renderer) too.
```

Commit:
fix: clean code

wattachai-lseg commented on documents/src/pages/elements/tree-select.md:
```diff
@@ -520,6 +520,48 @@ This control will save the current selection, and result of the firing of the `v
 
+## Manipulating item properties
+
+Item properties of Tree Select could be read and updated programmatically through its [TreeManager](./custom-components/utils/tree-manager) which is available as `treeManager` property. Retrieve [TreeNode(s)](./custom-components/utils/tree-node) representing each item by calling `getTreeNode()` or `getTreeNodes()` of `treeManager`. In adiition, they could be used for [custom renderer](./custom-co ...
```
```suggestion
Item properties of Tree Select could be read and updated programmatically through its [TreeManager](./custom-components/utils/tree-manager) which is available as `treeManager` property. Retrieve [TreeNode(s)](./custom-components/utils/tree-node) representing each item by calling `getTreeNode()` or `getTreeNodes()` of `treeManager`. In addition, they could be used for [custom renderer](./custom-components/utils/tree-node#custom-renderer) too.
```

Theeraphat-Sorasetsakul commented on scripts/release/utils.injector.js:
```diff
@@ -259,12 +272,14 @@ const generateMD = () => {
 
       const outputFile = path.resolve(Build.PAGES_FOLDER, output);
       if (fs.existsSync(outputFile)) {
+        // add new line to split with exist content.
```
```suggestion
        // add new line to split with existing content.
```

Theeraphat-Sorasetsakul committed:
Apply suggestions from code review

Co-authored-by: wattachai <117723407+wattachai-lseg@users.noreply.github.com>

wattachai-lseg committed:
feat: add depth param to get descendants of Tree Node

wattachai-lseg committed:
docs(tree): add missing JSDoc of Tree Node

dtanp-rft commented on documents/src/pages/elements/tree.md:
```diff
@@ -455,6 +455,50 @@ tree?.addEventListener('value-changed', (event) => {
 
+## Manipulating item properties
+
+Item properties of Tree could be read and updated programmatically through its [TreeManager](./custom-components/utils/tree-manager) which is available as `manager` property. Retrieve [TreeNode(s)](./custom-components/utils/tree-node) representing each item by calling `getTreeNode()` or `getTreeNodes()` of `manager`. In addition, they could be used for [custom renderer](./custom-components/utils/ ...
```
`In addition, they could be used for [custom renderer](./custom-components/utils/tree-node#custom-renderer) too.`

I feel that this content should be removed from this section. What we could do is to introduce another section, 'Custom Renderer' so we can provide completed guideline.

wattachai-lseg replied:
Let's move "Custom Renderer" from "Tree Node" to "Tree" & "Tree Select" + revise it content a little as readers focus on Component first rather than "Tree Node" first.

Theeraphat-Sorasetsakul replied:
Moved and wrote new Custom renderer section on both components.

dtanp-rft commented on documents/src/pages/elements/tree-select.md:
```diff
@@ -520,6 +520,48 @@ This control will save the current selection, and result of the firing of the `v
 
+## Manipulating item properties
+
+Item properties of Tree Select could be read and updated programmatically through its [TreeManager](./custom-components/utils/tree-manager) which is available as `treeManager` property. Retrieve [TreeNode(s)](./custom-components/utils/tree-node) representing each item by calling `getTreeNode()` or `getTreeNodes()` of `treeManager`. In addition, they could be used for [custom renderer](./custom-co ...
```
`In addition, they could be used for [custom renderer](./custom-components/utils/tree-node#custom-renderer) too.`

I feel that this content should be removed from this section. What we could do is to introduce another section, 'Custom Renderer' so we can provide completed guideline.

wattachai-lseg replied:
Let's move "Custom Renderer" from "Tree Node" to "Tree" & "Tree Select" + revise it content a little as readers focus on Component first rather than "Tree Node" first.

Theeraphat-Sorasetsakul replied:
Moved and wrote new Custom renderer section on both components.

dtanp-rft commented on packages/elements/src/tree/managers/tree-manager.ts:
```diff
@@ -40,6 +40,8 @@ export class TreeManager<T extends TreeDataItem> {
+   * Most of the time, there is no need to create a new instance of Tree Manager manually.
+   * Use the existing instance in components instead.
    * @param input Items or CollectionComposer to be managed.
    * @param mode TreeManager mode which is Relational or Independent.
```
```suggestion
   * @param mode Selected state of parent-children node has relationship or no-relationship.
```

wattachai-lseg replied:
Some additional options
* How items are managed either relationally or independently.
* How items' selected & checked state are managed either relationally or independently.

Theeraphat-Sorasetsakul replied:
Fixed to be `The mode describes how items are managed either relationally or independently.`

wattachai-lseg replied:
`a mode describing how items are managed either relationally or independently`

Theeraphat-Sorasetsakul replied:
Fixed.

dtanp-rft commented on packages/elements/src/tree/managers/tree-manager.ts:
```diff
@@ -134,8 +136,8 @@ export class TreeManager<T extends TreeDataItem> {
-   * Items which should be visibly displayed.
-   * This can be used to render items.
+   * Returns items which their selected state can be changed.
+   * hidden, disabled or readonly items are not included.
```
```suggestion
   * Hidden, disabled or readonly items are not included.
```

dtanp-rft commented on packages/elements/src/tree/managers/tree-manager.ts:
```diff
@@ -160,8 +162,8 @@ export class TreeManager<T extends TreeDataItem> {
-   * Items which should be visibly displayed.
-   * This can be used to render items.
+   * Returns currently displayed items.
+   * hidden and children of unexpanded items are not included.
```
```suggestion
   * Hidden and children of unexpanded items are not included.
```

dtanp-rft commented on packages/elements/src/tree/managers/tree-manager.ts:
```diff
@@ -494,7 +497,7 @@ export class TreeManager<T extends TreeDataItem> {
-   * Toggles the item checked state
+   * Toggle the selected state of the item.
    * @param item Original data item
    * @returns `True` if the item is modified
```
```suggestion
   * @returns `true` if the item is modified successfully.
```

dtanp-rft commented on packages/elements/src/tree/managers/tree-manager.ts:
```diff
@@ -430,23 +433,23 @@ export class TreeManager<T extends TreeDataItem> {
 
   /**
-   * Collapses all items in the tree
+   * Collapses all items managed by the Tree Manager.
```
Why it's necessary to mention 'managed by the Tree Manager'? I'd assume that there are possible that some items are not managed by Tree Manager and those items won't be collapsed. Is that right?

dtanp-rft commented on packages/elements/src/tree/managers/tree-manager.ts:
```diff
@@ -430,23 +433,23 @@ export class TreeManager<T extends TreeDataItem> {
 
   /**
-   * Expands all items in the tree
+   * Expands all items managed by the Tree Manager.
```
Why it's necessary to mention 'managed by the Tree Manager'? I'd assume that there are possible that some items are not managed by Tree Manager and those items won't be collapsed. Is that right?

wattachai-lseg replied:
There is no such case: all items must be managed by Tree Manager.  Is `Expands all items` too short?

dtanp-rft replied:
Looks OK to me. Concise and clear.

dtanp-rft commented on packages/elements/src/tree/managers/tree-manager.ts:
```diff
@@ -356,9 +359,9 @@ export class TreeManager<T extends TreeDataItem> {
+   * Return checked state of the item.
    * @param item Original data item
-   * @returns Checked state of the item
+   * @returns item checked state: `1` for checked, `0` for unchecked and `-1` for indeterminate
```
this looks a bit odd but I guess the intention is to focus on Javascript developer because TS developers would use CheckedState type so it will return 'CHECKED', 'INDETERMINATE', 'UNCHECKED'. Right? I might suggest to revise it to more generic statement. For example, `@returns item checked state: checked, unchecked or indeterminate.`

wattachai-lseg replied:
Without the actual number, JavaScript users would have no idea what these value are.
On the other hand, TypeScript users could easily map these 3 numbers back to their enum values.

dtanp-rft replied:
Yeah I see what you mean. Not strong feedback of mine here though. Happy to keep the same wording, my point is just make this more generic so it isn't too specific to any JS or TS but I agree that it could make the doc less useful too.

Theeraphat-Sorasetsakul replied:
Fixed.

Theeraphat-Sorasetsakul replied:
Fixed to be `item checked state: CHECKED (1), UNCHECKED (0), INDETERMINATE (-1)`

dtanp-rft commented on packages/elements/src/tree/managers/tree-manager.ts:
```diff
@@ -371,26 +374,26 @@ export class TreeManager<T extends TreeDataItem> {
-   * Gets an item's descendants
+   * Returns all descendants of the item.
    * @param item Original data item
    * @param depth Depth to retrieve
```
```suggestion
   * @param depth Specified depth of descendants to be retrieved.
```

dtanp-rft commented on packages/elements/src/tree/managers/tree-manager.ts:
```diff
@@ -329,7 +332,7 @@ export class TreeManager<T extends TreeDataItem> {
-   * Is the item expanded?
+   * Returns the current expanded state of the item.
    * @param item Original data item
    * @returns `True` if the item is expanded and its children should be visible
```
```suggestion
   * @returns `True` if the item is currently expanded so its children are visible.
```

dtanp-rft commented on packages/elements/src/tree/managers/tree-manager.ts:
```diff
@@ -284,7 +286,8 @@ export class TreeManager<T extends TreeDataItem> {
   /**
-   * Updates the data item, forcing a render
+   * Requests the item to be rerendered manually.
+   * This should be unnecessary most of the time as rerender is internally triggered by property update already.
```
```suggestion
   * Typically, this is not required. The render is triggered automatically when item's property are updated.
```

dtanp-rft reviewed (changes requested):
Accessors section in Tree Node looks a bit odd to me. There is no description on **Returns** but only **Type**.  it seems this whole section in not Accessors but Properties. Right?

I think it'd be good to have some examples at the top of Tree Manager page, similar to what we have in Tree Node. 

For simple example like 'checkItem', I am not sure from the doc what Argument that I should pass to this method as it said 'Original data item'. I don't have a clue what it is :)

wattachai-lseg committed:
docs: apply suggestions from code review

Co-authored-by: Domrongpon Tanpaibul <81604092+dtanp-rft@users.noreply.github.com>

wattachai-lseg commented:
> Accessors section in Tree Node looks a bit odd to me. There is no description on **Returns** but only **Type**. it seems this whole section in not Accessors but Properties. Right?

They are accessors working like properties similar to our components e.g. value accessor which apps could simply `textField.value += 'happy coding'`
 
> I think it'd be good to have some examples at the top of Tree Manager page, similar to what we have in Tree Node.

> For simple example like 'checkItem', I am not sure from the doc what Argument that I should pass to this method as it said 'Original data item'. I don't have a clue what it is :)

we could add the following to address both points as a new paragraph in Tree Manager doc page:

*Original data item refer to items passed to Tree Manager through `data` property of Tree & Tree-Select component*
*\<use code block to explain this with sample usage\>*

wattachai-lseg commented:
@Theeraphat-Sorasetsakul  would work on the feedbacks.

Theeraphat-Sorasetsakul commented on packages/elements/src/tree/managers/tree-manager.ts:
```diff
@@ -430,23 +433,23 @@ export class TreeManager<T extends TreeDataItem> {
 
   /**
-   * Expands all items in the tree
+   * Expands all items managed by the Tree Manager.
```
```suggestion
   * Expands all items.
```

Theeraphat-Sorasetsakul committed:
Update packages/elements/src/tree/managers/tree-manager.ts

Commit:
fix: rewrite some APIs description, fix accesser template to property

Sakchai-Refinitiv commented on documents/src/pages/utils/tree-manager.md:
```diff
@@ -0,0 +1,50 @@
+// Select the first item in Tree.
+const tree = document.querySelector('ef-tree');
+const data = [ 
+  { label: 'one', value: '1'}, 
```
```suggestion
  { label: 'one', value: '1' }, 
```

Theeraphat-Sorasetsakul replied:
Fixed

Sakchai-Refinitiv commented on documents/src/pages/utils/tree-manager.md:
```diff
@@ -0,0 +1,50 @@
+const data = [ 
+  { label: 'one', value: '1'}, 
+  { label: 'two', value: '2', icon: 'info' }
+];
```
The trailing spaces of each line of code make a bad DX when a developer copy and ESLint show warning message

Theeraphat-Sorasetsakul replied:
Fixed.

Sakchai-Refinitiv commented on documents/src/pages/utils/tree-node.md:
Expected output in the demo scripts shown `undefined` because a component is not ready which make a developer confusing especially a junior developer. you should find a way to improve DX

Sakchai-Refinitiv replied:
The expect output of line 46 must be ['3','2','4'] and we can mention about `selectedAt` sequential selection.

Sakchai-Refinitiv replied:
On the line 45, the arrow function should not return assignment to avoid a linting on developer IDE show warning/error to improve DX

Theeraphat-Sorasetsakul replied:
> Expected output in the demo scripts shown `undefined` because a component is not ready which make a developer confusing especially a junior developer. you should find a way to improve DX

As we talked, it's our known issue that it's rare care if the element is undefined.

> The expect output of line 46 must be ['3','2','4'] and we can mention about `selectedAt` sequential selection.

Fixed.

> On the line 45, the arrow function should not return assignment to avoid a linting on developer IDE show warning/error to improve DX

Fixed.

Commit:
fix: fix from feedback

Theeraphat-Sorasetsakul commented:
> Accessors section in Tree Node looks a bit odd to me. There is no description on **Returns** but only **Type**. it seems this whole section in not Accessors but Properties. Right?

Updated to change 'Accessors' header to 'Properties'. And also updated its style.

> I think it'd be good to have some examples at the top of Tree Manager page, similar to what we have in Tree Node.
> For simple example like 'checkItem', I am not sure from the doc what Argument that I should pass to this method as it said 'Original data item'. I don't have a clue what it is :)

Added. You will see an example of using 'checkItem' and suggestion if using TreeNode.

Theeraphat-Sorasetsakul committed:
Merge branch 'feat/tree-node/main' into feat/tree-node/doc

wattachai-lseg commented on documents/src/pages/elements/tree-select.md:
```diff
@@ -520,6 +520,155 @@ This control will save the current selection, and result of the firing of the `v
 
+## Manipulating item properties
+
+Item properties of Tree Select could be read and updated programmatically through its [TreeManager](./custom-components/utils/tree-manager) which is available as `treeManager` property. Retrieve [TreeNode(s)](./custom-components/utils/tree-node) representing each item by calling `getTreeNode()` or `getTreeNodes()` of `treeManager`.
```
```suggestion
Item properties of Tree Select could be read and updated programmatically through its [Tree Manager](./custom-components/utils/tree-manager) which is available as `treeManager` property. Retrieve [TreeNode(s)](./custom-components/utils/tree-node) representing each item by calling `getTreeNode()` or `getTreeNodes()` of `treeManager`.
```

wattachai-lseg commented on documents/src/pages/elements/tree-select.md:
```diff
@@ -520,6 +520,155 @@ This control will save the current selection, and result of the firing of the `v
+
+const treeNodes = treeSelect.treeManager.getTreeNodes();
+const selectedNode = treeNodes.find(treeNode => treeNode.value === 'ADR');
+selectedNode.selected = true; // Current output: []
```
```suggestion
selectedNode.selected = true;
```

wattachai-lseg commented on documents/src/pages/elements/tree-select.md:
```diff
@@ -520,6 +520,155 @@ This control will save the current selection, and result of the firing of the `v
+
+const treeNodes = treeSelect.treeManager.getTreeNodes();
+const selectedNode = treeNodes.find(treeNode => treeNode.value === 'ADR');
+selectedNode.selected = true; // Current output: []
```
```suggestion
selectedNode.selected = true;
```

wattachai-lseg commented on documents/src/pages/elements/tree-select.md:
```diff
@@ -520,6 +520,155 @@ This control will save the current selection, and result of the firing of the `v
+
+## Custom renderer
+
+TreeSelect supports custom rendering by providing a renderer property that can define how each item is displayed. The renderer is a callback function that takes a data item, a Collection Composer, and previously mapped item elements (if any), and must return an HTMLElement. You can use TreeNode to provides access to the current item's data property. Note that for performance sensitive use cases s ...
```
```suggestion
Tree Select defines how each of its item is displayed with `renderer` property. You can customise this renderer by setting a callback function to the property. [Tree Node](/custom-components/utils/tree-node) is the easiest way to implement the function. Note that for performance sensitive use cases such as a large number of items, consider using [Collection Composer](/custom-components/utils/data-management#collection-composer) instead.
```

wattachai-lseg commented on documents/src/pages/elements/tree-select.md:
```diff
@@ -520,6 +520,155 @@ This control will save the current selection, and result of the firing of the `v
+import { uuid } from '@refinitiv-ui/utils/uuid.js';
+import { CheckedState, TreeManager, TreeManagerMode } from '@refinitiv-ui/elements/tree';
+
+// Implement TreeSelect's default render with Tree Node instead of Collection Composer
```
```suggestion
// Implement Tree Select's default renderer with Tree Node instead of Collection Composer
```

wattachai-lseg commented on documents/src/pages/elements/tree-select.md:
```diff
@@ -520,6 +520,155 @@ This control will save the current selection, and result of the firing of the `v
+  treeManager?: TreeManager<T>;
+};
+
+// Implement TreeSelect's default render with Tree Node instead of Collection Composer
```
```suggestion
// Implement Tree Select's default renderer with Tree Node instead of Collection Composer
```

wattachai-lseg commented on documents/src/pages/elements/tree-select.md:
```diff
@@ -520,6 +520,155 @@ This control will save the current selection, and result of the firing of the `v
+    const multiple = _context?.multiple === true;
+    const noRelation = _context?.noRelation === true;
+    const mode = !multiple || !noRelation ? TreeManagerMode.RELATIONAL : TreeManagerMode.INDEPENDENT;
+    const manager = _context?.treeManager || new TreeManager(composer, mode);
```
```suggestion
    const manager = _context?.manager || _context?.treeManager || new TreeManager(composer, mode);
```

wattachai-lseg commented on documents/src/pages/elements/tree-select.md:
```diff
@@ -520,6 +520,155 @@ This control will save the current selection, and result of the firing of the `v
+    return _element;
+  };
+};
+treeSelect.renderer = createTreeRenderer(this)
```
```suggestion
treeSelect.renderer = createTreeRenderer(treeSelect)
```

wattachai-lseg commented on documents/src/pages/elements/tree-select.md:
```diff
@@ -520,6 +520,155 @@ This control will save the current selection, and result of the firing of the `v
+    return element;
+  };
+};
+treeSelect.renderer = createTreeRenderer(this)
```
```suggestion
treeSelect.renderer = createTreeRenderer(treeSelect)
```

wattachai-lseg commented on documents/src/pages/elements/tree.md:
````diff
@@ -455,10 +455,157 @@ tree?.addEventListener('value-changed', (event) => {
+```typescript
+// Select the item which value is '1.1'
+const tree = document.querySelector('ef-tree')!;
+tree.data = [
````
`data` would be an unknown property without proper type def. Missing `import type { Tree } from '@refinitiv-ui/elements/tree'`?

Theeraphat-Sorasetsakul replied:
Fixed.

wattachai-lseg commented on documents/src/pages/elements/tree-select.md:
````diff
@@ -520,6 +520,155 @@ This control will save the current selection, and result of the firing of the `v
+```typescript
+// Select the item which value is 'ADR'
+const treeSelect = document.querySelector('ef-tree-select')!;
+treeSelect.data = [
````
`data` would be an unknown property without proper type def. Missing import type { TreeSelect } from '@refinitiv-ui/elements/tree-select'?

Theeraphat-Sorasetsakul replied:
Fixed.

wattachai-lseg commented on documents/src/pages/elements/tree-select.md:
```diff
@@ -520,6 +520,155 @@ This control will save the current selection, and result of the firing of the `v
+console.log(treeSelect.values); // Expected output: []
+
+const treeNodes = treeSelect.treeManager.getTreeNodes();
+const selectedNode = treeNodes.find(treeNode => treeNode.value === 'ADR');
```
`selectedNode` is rather misleading: the node is not selected yet.
```suggestion
const node = treeNodes.find(treeNode => treeNode.value === 'ADR');
```

wattachai-lseg commented on documents/src/pages/elements/tree-select.md:
```diff
@@ -520,6 +520,155 @@ This control will save the current selection, and result of the firing of the `v
+console.log(treeSelect.values); // Expected output: []
+
+const treeNodes = treeSelect.treeManager.getTreeNodes();
+const selectedNode = treeNodes.find(treeNode => treeNode.value === 'ADR');
```
`selectedNode` is rather misleading: the node is not selected yet.
```suggestion
const node = treeNodes.find(treeNode => treeNode.value === 'ADR');
```

wattachai-lseg commented on documents/src/pages/elements/tree.md:
```diff
@@ -455,10 +455,157 @@ tree?.addEventListener('value-changed', (event) => {
+console.log(tree.values); // Expected output: []
+
+const treeNodes = tree.manager.getTreeNodes();
+const selectedNode = treeNodes.find(treeNode => treeNode.value === '1.1');
```
`selectedNode` is rather misleading: the node is not selected yet.
```suggestion
const node = treeNodes.find(treeNode => treeNode.value === '1.1');
```

wattachai-lseg commented on documents/src/pages/elements/tree.md:
```diff
@@ -455,10 +455,157 @@ tree?.addEventListener('value-changed', (event) => {
+console.log(tree.values); // Expected output: []
+
+const treeNodes = tree.manager.getTreeNodes();
+const selectedNode = treeNodes.find(treeNode => treeNode.value === '1.1');
```
`selectedNode` is rather misleading: the node is not selected yet.
```suggestion
const node = treeNodes.find(treeNode => treeNode.value === '1.1');
```

wattachai-lseg commented on documents/src/pages/elements/tree.md:
```diff
@@ -455,10 +455,157 @@ tree?.addEventListener('value-changed', (event) => {
 
+## Manipulating item properties
+
+Item properties of Tree could be read and updated programmatically through its [TreeManager](./custom-components/utils/tree-manager) which is available as `manager` property. Retrieve [TreeNode(s)](./custom-components/utils/tree-node) representing each item by calling `getTreeNode()` or `getTreeNodes()` of `manager`.
```
```suggestion
Item properties of Tree could be read and updated programmatically through its [Tree Manager](./custom-components/utils/tree-manager) which is available as `manager` property. Retrieve [TreeNode(s)](./custom-components/utils/tree-node) representing each item by calling `getTreeNode()` or `getTreeNodes()` of `manager`.
```

wattachai-lseg commented on documents/src/pages/elements/tree.md:
```diff
@@ -455,10 +455,157 @@ tree?.addEventListener('value-changed', (event) => {
+
+## Custom renderer
+
+Tree supports custom rendering by providing a renderer property that can define how each item is displayed. The renderer is a callback function that takes a data item, a Collection Composer, and previously mapped item elements (if any), and must return an HTMLElement. You can use TreeNode to provides access to the current item's data property. Note that for performance sensitive use cases such as ...
```
```suggestion
Tree defines how each of its item is displayed with `renderer` property. You can customise this renderer by setting a callback function to the property. [Tree Node](/custom-components/utils/tree-node) is the easiest way to implement the function. Note that for performance sensitive use cases such as a large number of items, consider using [Collection Composer](/custom-components/utils/data-management#collection-composer) instead.
```

Theeraphat-Sorasetsakul commented on packages/elements/src/tree/managers/tree-manager.ts:
```diff
@@ -44,17 +44,19 @@ export class TreeManager<T extends TreeDataItem> {
+   * Use the existing instance in components instead.
    * @param input Items or CollectionComposer to be managed.
-   * @param mode TreeManager mode which is Relational or Independent.
+   * @param mode The mode describes how items are managed either relationally or independently.
```
```suggestion
   * @param mode A mode describing how items are managed either relationally or independently.
```

Theeraphat-Sorasetsakul committed:
Update packages/elements/src/tree/managers/tree-manager.ts

wattachai-lseg commented on documents/src/pages/elements/tree.md:
```diff
@@ -455,10 +455,157 @@ tree?.addEventListener('value-changed', (event) => {
+import { uuid } from '@refinitiv-ui/utils/uuid.js';
+import { CheckedState, TreeManager, TreeManagerMode } from '@refinitiv-ui/elements/tree';
+
+// Implement Tree's default render with Tree Node instead of Collection Composer
```
```suggestion
// Implement Tree's default renderer with Tree Node instead of Collection Composer
```

wattachai-lseg commented on documents/src/pages/elements/tree.md:
```diff
@@ -455,10 +455,157 @@ tree?.addEventListener('value-changed', (event) => {
+    return element;
+  };
+};
+tree.renderer = createTreeRenderer(this)
```
```suggestion
tree.renderer = createTreeRenderer(tree)
```

wattachai-lseg commented on documents/src/pages/elements/tree.md:
```diff
@@ -455,10 +455,157 @@ tree?.addEventListener('value-changed', (event) => {
+    return _element;
+  };
+};
+tree.renderer = createTreeRenderer(this)
```
```suggestion
tree.renderer = createTreeRenderer(tree)
```

wattachai-lseg commented on documents/src/pages/elements/tree.md:
```diff
@@ -455,10 +455,157 @@ tree?.addEventListener('value-changed', (event) => {
+  treeManager?: TreeManager<T>;
+};
+
+// Implement Tree's default render with Tree Node instead of Collection Composer
```
```suggestion
// Implement Tree's default renderer with Tree Node instead of Collection Composer
```

wattachai-lseg commented on documents/src/pages/elements/tree.md:
```diff
@@ -455,10 +455,157 @@ tree?.addEventListener('value-changed', (event) => {
+    const multiple = _context?.multiple === true;
+    const noRelation = _context?.noRelation === true;
+    const mode = !multiple || !noRelation ? TreeManagerMode.RELATIONAL : TreeManagerMode.INDEPENDENT;
+    const manager = _context?.manager || new TreeManager(composer, mode);
```
```suggestion
    const manager = _context?.manager || _context?.treeManager || new TreeManager(composer, mode);
```

wattachai-lseg commented on documents/src/pages/utils/tree-manager.md:
````diff
@@ -0,0 +1,50 @@
+console.log(tree.values); // Expected output: ['1']
+```
+
+Alternatively, [Tree Node](/custom-components/utils/tree-node) is the suggested way to work with Tree Manager. Most of the time, getTreeNodes(), getTreeNode() and TreeNode are all you need.
````
```suggestion
Alternatively, [Tree Node](/custom-components/utils/tree-node) is the suggested way to manage items in Tree Manager. Along with `getTreeNodes()` and `getTreeNode()`, they are all your need for most use cases.
```

wattachai-lseg commented on documents/src/pages/utils/tree-manager.md:
````diff
@@ -0,0 +1,50 @@
+firstItem.selected = true
+```
+
+@> Tree Manager supports type generic defining the type of its items. In this document, this generic type is noted as `T`.
````
```suggestion
@> Tree Manager supports type generic defining the type of its items. In this document, this generic type is denoted as `T`.
```

wattachai-lseg commented on documents/src/pages/utils/tree-node.md:
```diff
@@ -13,135 +13,39 @@ Tree Node provides an easy way to programmatically manipulate items in [Tree](/e
 console.log(tree.values); // Expected output: ['3']
 
 const treeNodes = tree.manager.getTreeNodes();
 const selectedNodes = treeNodes.filter(treeNode => treeNode.icon);
```
```suggestion
const targetNodes = treeNodes.filter(treeNode => treeNode.icon);
```

wattachai-lseg commented on documents/src/pages/utils/tree-node.md:
```diff
@@ -13,135 +13,39 @@ Tree Node provides an easy way to programmatically manipulate items in [Tree](/e
 console.log(tree.values); // Expected output: ['3']
 
 const treeNodes = tree.manager.getTreeNodes();
 const selectedNodes = treeNodes.filter(treeNode => treeNode.icon);
```
```suggestion
const targetNodes = treeNodes.filter(treeNode => treeNode.icon);
```

wattachai-lseg commented on packages/elements/src/tree/managers/tree-manager.ts:
```diff
@@ -70,10 +72,10 @@ export class TreeManager<T extends TreeDataItem> {
   /**
-   * Return a `TreeNode` of the specified item.
+   * Returns a `TreeNode` of the original data item.
    * If the item doesn't exist, return `null`.
```
```suggestion
   * If the item doesn't exist, returns `null`.
```

wattachai-lseg commented on packages/elements/src/tree/managers/tree-manager.ts:
```diff
@@ -288,7 +290,8 @@ export class TreeManager<T extends TreeDataItem> {
   /**
-   * Updates the data item, forcing a render
+   * Requests the item to be rerendered manually.
+   * Typically, this is not required. The render is triggered automatically when item's property are updated.
```
```suggestion
   * Typically, this is not required. The render is triggered automatically when item's properties are updated.
```

wattachai-lseg commented on packages/elements/src/tree/managers/tree-manager.ts:
```diff
@@ -375,26 +378,26 @@ export class TreeManager<T extends TreeDataItem> {
    * @param item Original data item
-   * @param depth Depth to retrieve
-   * @returns A list of descendants
+   * @param depth Specified depth of descendants to be retrieved.
```
```suggestion
   * @param depth Depth of descendants to get. If it's `undefined`, get all descendants.
```

wattachai-lseg commented on packages/elements/src/tree/managers/tree-node.ts:
```diff
@@ -145,40 +145,69 @@ export class TreeNode<T extends TreeDataItem = TreeDataItem> {
-    const descendants = this.manager.getItemDescendants(this.item);
+  /**
+   * Returns all descendants of the item.
+   * @param depth Depth to retrieve
```
```suggestion
   * @param depth Depth of descendants to get. If it's `undefined`, get all descendants.
```

wattachai-lseg commented on packages/elements/src/tree/managers/tree-node.ts:
```diff
@@ -145,40 +145,69 @@ export class TreeNode<T extends TreeDataItem = TreeDataItem> {
+   */
   public isChild(): boolean {
     return this.manager.isItemChild(this.item);
   }
```
Let's update `rerender` JSDoc desciption to the same one with Tre Manager.

> Requests the item to be rerendered manually. Typically, this is not required. The render is triggered automatically when item's properties are updated.

Theeraphat-Sorasetsakul replied:
Updated.

wattachai-lseg reviewed (changes requested):
This should be everything

Theeraphat-Sorasetsakul committed:
Apply suggestions from code review

Co-authored-by: wattachai <117723407+wattachai-lseg@users.noreply.github.com>

Commit:
fix: fix example in docs

Commit:
fix: fix missing import type in example in docs

wattachai-lseg committed:
Merge branch 'feat/tree-node/main' into feat/tree-node/doc
